scikeras/_saving_utils.py

```diff
@@ -0,0 +1,140 @@
+    """A hack to restore weights in optimizers that use slots.
+
+    See https://tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer#slots_2
+    """
+    self._create_all_weights_orig(var_list)
+    try:
+        self.set_weights(self._restored_weights)
+    except ValueError:
+        # Weights don't match, eg. when optimizer was pickled before any training
+        # or a completely new dataset is being used right after pickling
+        pass
+    delattr(self, "_restored_weights")
+    self._create_all_weights = self._create_all_weights_orig
+
+
+def _restore_optimizer_weights(optimizer, weights) -> None:
+    optimizer._restored_weights = weights
+    optimizer._create_all_weights_orig = optimizer._create_all_weights
+    # MethodType is used to "bind" the _temp_create_all_weights method
+    # to the "live" optimizer object
+    optimizer._create_all_weights = MethodType(_temp_create_all_weights, optimizer)
+
+
+def unpack_keras_model(packed_keras_model, optimizer_weights):
+    """Reconstruct a model from the result of __reduce__
+    """
+    temp_dir = _get_temp_folder()
+    b = BytesIO(packed_keras_model)
+    with zipfile.ZipFile(b, "r", zipfile.ZIP_DEFLATED) as zf:
+        for path in zf.namelist():
+            dest = os.path.join(temp_dir, path)
+            tf_io.gfile.makedirs(os.path.dirname(dest))
+            with tf_io.gfile.GFile(dest, "wb") as f:
+                f.write(zf.read(path))
+    model: keras.Model = load_model(temp_dir)
+    for root, _, filenames in tf_io.gfile.walk(temp_dir):
+        for filename in filenames:
+            if filename.startswith("ram://"):
+                # Currently, tf.io.gfile.walk returns
+                # the entire path for the ram:// filesystem
+                dest = filename
+            else:
+                dest = os.path.join(root, filename)
+            tf_io.gfile.remove(dest)
+    _restore_optimizer_weights(model.optimizer, optimizer_weights)
+    return model
+
+
+def pack_keras_model(model):
+    """Support for Pythons's Pickle protocol.
+    """
+    temp_dir = _get_temp_folder()
+    model.save(temp_dir)
+    b = BytesIO()
+    with zipfile.ZipFile(b, "w", zipfile.ZIP_DEFLATED) as zf:
+        for root, _, filenames in tf_io.gfile.walk(temp_dir):
+            for filename in filenames:
+                if filename.startswith("ram://"):
+                    # Currently, tf.io.gfile.walk returns
+                    # the entire path for the ram:// filesystem
+                    dest = filename
+                else:
+                    dest = os.path.join(root, filename)
+                with tf_io.gfile.GFile(dest, "rb") as f:
+                    zf.writestr(os.path.relpath(dest, temp_dir), f.read())
+                tf_io.gfile.remove(dest)
+    b.seek(0)
+    return (
+        unpack_keras_model,
+        (np.asarray(memoryview(b.read())), model.optimizer.get_weights()),
+    )
+
+
+def unpack_keras_optimizer(opt_serialized, weights):
```

Comment:
Why does can't `_create_all_weights` be a regular function? It'd be a little cleaner if this all moved to the `Optimizer` class.

<details>
<summary>Draft implementation</summary>
I think this is a little cleaner; some functions look like they should be class methods. This implementation might be missing some (possibly important) details.

``` python
def _unpack_optimizer(opt_serialized, weights):
    opt = keras.optimizers.deserialize(opt_serialized)
    opt._restore_weights(weights)
    return opt

class Optimizer(...):
    def __reduce__(self):
        weights = self.get_weights()
        opt_serialized = keras.optimizers.serialize(optimizer)
        return unpack_keras_optimizer, (opt_serialized, weights)

    def _restore_weights(self, weights):
        self._restored_weights = weights
        self._create_all_weights_orig = self._create_all_weights
        self._create_all_weights = MethodType(_temp_create_all_weights, self)
```

</details>

Reply:
I may be missing something, so sorry if this causes confusion:

> Why can't _create_all_weights be a regular function?

I'm not sure what you mean by "a regular function"? `Optimizer._create_all_weights` is a function, `_temp_create_all_weights` is a function that replaces `Optimizer._create_all_weights` so that weights get restored after `Optimizer._create_all_weights` is called for the first time, and that then cleans up the extra weights and un monkey patches itself.

> It'd be a little cleaner if this all moved to the Optimizer class

We can't directly edit the `Optimizer` class, only monkey patch methods onto it. I'm not sure what the advantage of monkey patching `_restore_weights` onto `Optimizer` is vs. referencing it directly.

We also need to call `_restore_optimizer_weights` from `unpack_keras_model` because the implementation is not recursive: when optimizers are serialized as part of a model, `Optimizer.__reduce__` is never called.